### PR TITLE
MariaDB

### PR DESCRIPTION
**Pull request by Iain-S**
Adds MariaDB support, which might also work for a wider array of DBs (sqlite perhaps?).

### DSNs

This PR is a breaking change since we now require users to provide the DB connection DSNs (e.g. `postgresql://user:secret@localhost:6789/dbname`) rather than building them from `SRC_DBNAME`, `SRC_HOST`, `SRC_PASSWORD`, etc.
I find this less convenient, as an end user, but I think it's safest as the DSN is all we really need to pass to SQLAlchemy and this change

1. works for mariadb and postgres
2. future proofs us against other dbms we want to support
3. allows users to specify weird options, like this two-server postgres dsn (not tested) `postgresql://host1:123,host2:456/somedb`

### Choosing providers

We modify the method of choosing a provider for a given column type. We now use a dict and try to `.get()` a match based on the column type and whether the column has a `length` attribute. If that fails, we iterate over all known column types and check `issubclass(our_column_type, dict_column_type)`. This should result in fewer `Unsupported type...` runtime errors.

### Loan tutorial

We add a short tutorial page to describe how to run SSG on [this](https://relational.fit.cvut.cz/dataset/Financial) dataset.

### SQLite3

Further to my speculation above, I have tested SSG with a simple SQLite3 DB using these env vars:

```text
src_dsn="sqlite:////Users/me/somedir/src.sqlite"
src_dsn="sqlite:////Users/me/somedir/dst.sqlite"
```

and it works!

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ jsonschema = "^4.17.3"
 sqlacodegen = "3.0.0rc1"
 asyncpg = "^0.27.0"
 greenlet = "^2.0.2"
+pymysql = "^1.1.0"
```
Comment (cptanalatriste) [inline pyproject.toml:27]: Maybe to future-proof SSG, we might need to add a way for dynamically adding a DB connector. In Java, you can drop JAR files containing the drivers and add them to the classpath (FYI, this is just a comment, not a change request :P )

File: sqlsynthgen/main.py [hunk 2]
```diff
@@ -35,12 +34,12 @@ def _check_file_non_existence(file_path: Path) -> None:
         sys.exit(1)
 
 
-def _get_src_postgres_dsn(settings: Settings) -> PostgresDsn:
+def _get_src_postgres_dsn(settings: Settings) -> str:
```
Comment (cptanalatriste) [inline sqlsynthgen/main.py:37]: Maybe we can rename this method? It appears its producing connections strings for other databases than PostgreSQL

File: sqlsynthgen/make.py [hunk 3]
```diff
@@ -210,33 +209,35 @@ def _get_mimesis_function_for_colum(column: Any) -> Tuple[List[str], str, List[s
...
+        (sqltypes.LargeBinary, False): "generic.bytes_provider.bytes",
+        (sqltypes.String, False): "generic.text.color",
+        (sqltypes.String, True): "generic.person.password",
+    }
+
+    generator_function = mapping.get((column_type, column_size is not None), None)
+
+    if not generator_function:
+        for key, value in mapping.items():
+            if issubclass(column_type, key[0]) and key[1] == (column_size is not None):
```
Comment (cptanalatriste) [inline sqlsynthgen/make.py:231]: I like this! Way better than my messy `if-elsif`

File: tests/test_make.py [hunk 4]
```diff
@@ -119,6 +126,54 @@ def test_make_generators_force_overwrite(
 
         self.assertEqual(expected, actual)
 
+    def test__get_provider_for_column(self) -> None:
```
Comment (cptanalatriste) [inline tests/test_make.py:129]: Maybe use only one underscore between `test` and `get_`?

**Review by cptanalatriste — APPROVED** [review]
Looks good to me! Only minor re-naming suggestions.

File: pyproject.toml
Comment (cptanalatriste) [inline pyproject.toml:1]: Also, do we need to change the `SSG` version to signal the breaking change?

File: sqlsynthgen/main.py (hunk 2 above)
Comment (Iain-S) [reply sqlsynthgen/main.py:37]: Good point, I've renamed it and also removed some other references to Postgres.

File: tests/test_make.py (hunk 4 above)
Comment (Iain-S) [reply tests/test_make.py:129]: Agreed. Technically, the function it is testing is called `_get_provider_for_column` but it doesn't read nicely.

File: pyproject.toml
Comment (Iain-S) [reply pyproject.toml:1]: Yes, except that we haven't actually done a release in a while so I think 0.3.0 is still suitable.

**Pull request merged**